File: node/lib/protocol.js
var Buffer = require('buffer').Buffer,
    stream = require('stream'),
    Class = require('js-class'),
    Empty = require('google-protobuf/google/protobuf/empty_pb.js').Empty,
    ErrorProto = require('../gen/tbus/error_pb.js').Error;

var PFX_ROUTING_MASK = 0xe0,
    PFX_ROUTING = 0xc0,
    PFX_ROUTING_ADDRNUM = 0x1f;

var FORMAT_MASK = 0xf0,
    FORMAT = 0x10;      // rev 1, protobuf encoded

var BF_ERROR = 0x80;    // body contains error

var EVT_ROUTE = 'route',
    EVT_HEAD  = 'head',
    EVT_MSG   = 'message';

var _logger = null;

function setLogger(logger) {
    _logger = logger;
}

var StreamDecoder = Class({
    constructor: function (initialState, options) {
        this._initialState = initialState;
        this._state = initialState;
        this._logger = _logger;
        this.options = options || {};
    },

    log: function (logger) {
        this._logger = logger;
        return this;
    },

    decode: function (buf) {
        this._buf = buf;
        while (this._result == null &&
            this._buf != null && this._buf.length > 0) {
            this._log('decode', function () {
                return {
                    state: this._state,
                    data: Array.from(this._buf, function (v) { return v.toString(16); }).join()
                }
            }.bind(this));
            this['_parse' + this._state]();
        }
        var result = this._result;
        delete this._result;
        return result;
    },

    readUInt8: function () {
        var b = this._buf.readUInt8(0);
        this.skipUInt8();
        return b;
    },

    decode7Bit: function () {
        var b = this.readUInt8();
        var v = (b & 0x7f) << (7 * (this._recvLen - 1));
        if (b & 0x80) {
            this._expLen ++;
            b &= 0x7f;
        }
        return v;
    },

    skipUInt8: function () {
        this._buf = this._buf.slice(1);
        this._recvLen ++;
    },

    transit: function (state) {
        this._log('transit', { from: this._state, to: state });
        this._state = state;
        this._expLen = 0;
        this._recvLen = 0;
        var enterFn = this['_enter' + state];
        if (enterFn != null) {
            enterFn.call(this);
        }
    },

    reset: function () {
        this.transit(this._initialState);
    },

    report: function (err, event, result) {
        this.setResult(err, event, result);
    },

    setResult: function (err, event, result) {
        this._result = {
            event: event,
            buf: this._buf.length > 0 ? this._buf : null
        };
        if (err != null) {
            this._result.err = err;
        } else {
            this._result.result = result;
        }
        this._log('report', this._result);
    },

    done: function (err, event, result) {
        this.report(err, event, result);
        this.reset();
    },

    allRecv: function () {
        return this._recvLen >= this._expLen;
    },

    _log: function (event, data) {
        if (this._logger) {
            if (typeof(data) == 'function') {
                data = data();
            }
            if (this.options.id != null) {
                event = this.options.id + ":" + event;
            }
            this._logger(event, data);
        }
    }
});

var Decoder = Class(StreamDecoder, {
    constructor: function (options) {
        StreamDecoder.prototype.constructor.call(this, 'Prefix', options);
        this._prefix = [];
    },

    _enterPrefix: function () {
        this._prefix = [];
    },

    _parsePrefix: function () {
        var b = this._buf.readUInt8(0);
        if ((b & PFX_ROUTING_MASK) == PFX_ROUTING) {
            this._addrNum = (b & PFX_ROUTING_ADDRNUM) + 1;
            this.skipUInt8();
            this._prefix.push(b);
            this.transit('Addrs');
        } else {
            this.transit('Flags');
        }
    },

    _enterAddrs: function () {
        this._expLen = this._addrNum;
    },

    _parseAddrs: function () {
        this._prefix.push(this.readUInt8());
        if (this.allRecv()) {
            var msg = {head: {prefix: this._prefix[0]}};
            msg.head.raw = new Buffer(this._prefix);
            msg.head.addrs = this._prefix.slice(1);
            msg.head.rawPrefix = msg.head.raw.slice(0);
            this.report(null, EVT_ROUTE, msg);
            this.transit('Flags');
        }
    },

    _enterFlags: function () {
        delete this._flag;
        delete this._msgId;
        delete this._bodyBytes;
        delete this._body;
        this._headBytes = 0;
        this._head = this._prefix.slice(0);
    },

    _parseFlags: function () {
        this._flag = this.readUInt8();
        this._head.push(this._flag);
        this._headBytes ++;
        if ((this._flag & FORMAT_MASK) != FORMAT) {
            this.done(new Error('unsupported protocol'));
        } else {
            this.transit('MsgId');
        }
    },

    _enterMsgId: function () {
        this._expLen = 1;
        this._msgId = 0;
    },

    _parseMsgId: function () {
        this._head.push(this._buf[0]);
        this._msgId |= this.decode7Bit();
        this._headBytes ++;
        if (this.allRecv()) {
            this.transit('BodySize');
        }
    },

    _enterBodySize: function () {
        this._expLen = 1;
        this._bodyBytes = 0;
        delete this._body;
    },

    _parseBodySize: function () {
        this._head.push(this._buf[0]);
        this._bodyBytes |= this.decode7Bit();
        this._headBytes ++;
        if (this.allRecv()) {
            this._report(EVT_HEAD);
            if (this._bodyBytes > 0) {
                this.transit('Body');
            } else {
                this.reset();
            }
        }
    },

    _enterBody: function () {
        this._expLen = this._bodyBytes;
        this._body = new Buffer(0);
    },

    _parseBody: function () {
        var size = this._buf.length;
        if (this._recvLen + size > this._expLen) {
            size = this._expLen - this._recvLen;
            this._body = Buffer.concat([this._body, this._buf.slice(0, size)]);
        } else {
            this._body = Buffer.concat([this._body, this._buf]);
        }
        this._buf = this._buf.slice(size);
        this._recvLen += size;
        if (this.allRecv()) {
            this._report(EVT_MSG);
            this.reset();
        }
    },

    _report: function (event) {
        var msg = {
            head: {
                flag: this._flag,
                msgId: this._msgId,
                bodyBytes: this._bodyBytes,
                raw: new Buffer(this._head)
            },
            body: {}
        };
        if (this._prefix.length > 0) {
            msg.head.prefix = this._prefix[0];
            msg.head.addrs = this._prefix.slice(1);
            msg.head.rawPrefix = msg.head.raw.slice(0, this._prefix.length);
            msg.head.rawHead = msg.head.raw.slice(this._prefix.length);
        } else {
            msg.head.rawHead = msg.head.raw.slice(0);
        }
        if (this._body != null && this._body.length > 0) {
            msg.body.raw = this._body;
            msg.body.flag = this._body.readUInt8(0);
            msg.body.data = this._body.slice(1);
        }
        this.report(null, event, msg);
    },
});

var DecodeStream = Class(stream.Writable, {
    constructor: function (decoder) {
        stream.Writable.call(this);
        this._decoder = decoder || new Decoder();
    },

    decoder: function () {
        return this._decoder;
    },

    close: function () {
        this.removeAllListeners();
    },

    _write: function (buf, encoding, callback) {
        while (buf != null && buf.length > 0) {
            var result = this._decoder.decode(buf);
            if (result == null) {
                break;
            }
            buf = result.buf;
            if (result.err != null) {
                this._handleErr(result.err)
            } else {
                this._handleEvent(result.event, result.result);
            }
        }
        setImmediate(callback);
    },

    _handleErr: function (err) {
        this.emit('error', err);
    },

    _handleEvent: function (event, msg) {
        setImmediate(function () {
            this.emit(event, msg);
        }.bind(this));
    }
});

function encode7Bit(bytes, value) {
    do {
        var v = value & 0x7f;
        value >>= 7;
        if (value > 0) {
            v |= 0x80;
        }
        bytes.push(v);
    } while (value > 0);
}

var Encoder = Class({
    constructor: function () {
        this._flag = FORMAT;
    },

    messageId: function (msgId) {
        this._msgId = msgId;
        return this;
    },

    body: function (body) {
        this._body = body;
        return this;
    },

    encodeError: function (err) {
        var reply = new ErrorProto();
        reply.setMessage(err.message);
        return this.encodeBody(BF_ERROR, reply.serializeBinary());
    },

    encodeProto: function (flag, pbMsg) {
        if (pbMsg == null) {
            pbMsg = new Empty();
        }
        return this.encodeBody(flag, pbMsg.serializeBinary());
    },

    encodeBody: function (flag, body) {
        if (!Buffer.isBuffer(body)) {
            body = new Buffer(body.buffer);
        }
        var buf = new Buffer(1 + body.length);
        buf.writeUInt8(flag, 0);
        body.copy(buf, 1);
        return this.body(buf);
    },

    route: function (addrs) {
        if (Array.isArray(addrs) && addrs.length > 0) {
            this._addrs = addrs;
        } else {
            delete(this._addrs);
        }
        return this;
    },

    buildMsg: function () {
        var msg = {
            head: {
                flag: this._flag,
                msgId: this._msgId
            },
            body: {
                raw: this._body
            }
        };
        var headBytes = 0;
        if (this._addrs && this._addrs.length > 0) {
            msg.head.addrs = this._addrs;
            headBytes += this._addrs.length + 1;
            var addrNum = (this._addrs.length - 1) & PFX_ROUTING_ADDRNUM;
            msg.head.prefix = PFX_ROUTING | addrNum;
        }
        var head = [this._flag];
        msg.head.bodyBytes = this._body ? this._body.length : 0;
        encode7Bit(head, this._msgId);
        encode7Bit(head, msg.head.bodyBytes);
        headBytes += head.length;

        msg.head.raw = new Buffer(headBytes);
        var off = 0;
        if (msg.head.addrs) {
            msg.head.raw.writeUInt8(msg.head.prefix, off++);
            for (var i = 0; i < msg.head.addrs.length; i ++) {
                msg.head.raw.writeUInt8(msg.head.addrs[i], off++);
            }
        }
        for (var i = 0; i < head.length; i ++) {
            msg.head.raw.writeUInt8(head[i], off++);
        }
        if (msg.head.addrs) {
            msg.head.rawPrefix = msg.head.raw.slice(0, msg.head.addrs.length + 1);
            msg.head.rawHead = msg.head.raw.slice(msg.head.addrs.length + 1);
        } else {
            msg.head.rawHead = msg.head.raw.slice(0);
        }

        if (msg.body.raw.length > 0) {
            msg.body.flag = msg.body.raw.readUInt8(0);
            msg.body.data = msg.body.raw.slice(1);
        }

        return msg;
    },

    toBuffer: function () {
        var msg = this.buildMsg();
        return Buffer.concat([msg.head.raw, msg.body.raw]);
    }
});

function decodeError(msg) {
    if (msg.body.flag & 0x80) {
        try {
            var err = ErrorProto.deserializeBinary(new Uint8Array(msg.body.data));
            return new Error(err.getMessage());
        } catch (err) {
            return err;
        }
    }
    return null;
}

function needRoute(msg) {
    return Array.isArray(msg.head.addrs) && msg.head.addrs.length > 0;
}

module.exports = {
    EVT_ROUTE:  EVT_ROUTE,
    EVT_HEAD:   EVT_HEAD,
    EVT_MSG:    EVT_MSG,

    Encoder: Encoder,
    Decoder: Decoder,
    DecodeStream: DecodeStream,

    decodeError: decodeError,
    needRoute: needRoute,
    setLogger: setLogger
};

File: go/tbus/protocol/protocol.go
package protocol

import (
	"bytes"
	"fmt"
	"io"
)

// Prefix and Flag
const (
	PfxRoutingMask    uint8 = 0xe0
	PfxRouting        uint8 = 0xc0
	PfxRoutingAddrNum uint8 = 0x1f

	RoutingAddrsMax = 32

	FormatMask uint8 = 0xf0
	Format     uint8 = 0x10 // rev 1, protobuf encoded

	BodyError uint8 = 0x80 // body contains error
)

// MsgHead is message head
type MsgHead struct {
	Prefix    uint8
	Addrs     []uint8
	Flag      uint8
	MsgID     uint32
	BodyBytes uint16
	Raw       []byte
	RawPrefix []byte
	RawHead   []byte
}

// NeedRoute indicates the message contains routing addresses
func (h *MsgHead) NeedRoute() bool {
	return (h.Prefix&PfxRoutingMask) == PfxRouting && len(h.Addrs) > 0
}

// MsgBody is message body
type MsgBody struct {
	Flag uint8
	Data []byte
	Raw  []byte
}

// Msg is a completely message
type Msg struct {
	Head MsgHead
	Body MsgBody
}

// DecodeHead decodes message head from input
func DecodeHead(reader io.Reader) (head MsgHead, err error) {
	var buf bytes.Buffer
	if _, err = io.CopyN(&buf, reader, 1); err != nil {
		return
	}
	pfx := uint8(buf.Bytes()[0])
	var addrNum int
	if (pfx & PfxRoutingMask) == PfxRouting {
		head.Prefix = pfx
		addrNum = int((pfx & PfxRoutingAddrNum) + 1)
		if _, err = io.CopyN(&buf, reader, int64(addrNum)); err != nil {
			return
		}
	} else if (pfx & FormatMask) == Format {
		head.Flag = pfx
		if head.MsgID, err = decode7bit32(&buf, reader); err != nil {
			return
		}
		if head.BodyBytes, err = decode7bit16(&buf, reader); err != nil {
			return
		}
	} else {
		err = fmt.Errorf("unknown byte 0x%02x", pfx)
		return
	}

	head.Raw = buf.Bytes()
	if addrNum > 0 {
		head.Addrs = head.Raw[1 : addrNum+1]
		head.RawPrefix = head.Raw[0 : addrNum+1]
		head.RawHead = head.Raw[addrNum+1:]
	} else {
		head.RawHead = head.Raw
	}
	return
}

// DecodeBody decodes message body from input
func DecodeBody(reader io.Reader, bodyBytes uint16) (body MsgBody, err error) {
	body.Raw = make([]byte, bodyBytes)
	if _, err = io.ReadFull(reader, body.Raw); err != nil {
		return
	}
	if bodyBytes > 0 {
		body.Flag = uint8(body.Raw[0])
		body.Data = body.Raw[1:]
	}
	return
}

// Decode decodes a complete message
func Decode(reader io.Reader) (msg Msg, err error) {
	if msg.Head, err = DecodeHead(reader); err != nil {
		return
	}
	msg.Body, err = DecodeBody(reader, msg.Head.BodyBytes)
	return
}

// DecodeError returns error if the message contains error
func DecodeError(msg *Msg) error {
	// TODO
	return nil
}

// EncodeRoute encodes routing prefix
func EncodeRoute(w io.Writer, addrs []uint8) (n int, err error) {
	if l := len(addrs); l > RoutingAddrsMax {
		panic("number of addresses exceeds limit")
	} else if l > 0 {
		pfx := []byte{PfxRouting | uint8(l-1)}
		if _, err = w.Write(pfx); err != nil {
			return
		}
		if n, err = w.Write(addrs); err != nil {
			n = 1
			return
		}
		n++
	}
	return
}

// EncodeMsg encodes a message
func EncodeMsg(w io.Writer, msgID uint32, bodyFlag uint8, body []byte) (n int, err error) {
	b := []byte{Format}
	if _, err = w.Write(b); err != nil {
		return
	}
	n++

	var c int
	if c, err = encode7bit(w, uint64(msgID)); err != nil {
		return
	}
	n += c
	if c, err = encode7bit(w, uint64(len(body)+1)); err != nil {
		return
	}
	n += c
	b[0] = bodyFlag
	if _, err = w.Write(b); err != nil {
		return
	}
	n++
	if c, err = w.Write(body); err != nil {
		return
	}
	n += c
	return
}

// Encode encodes routing prefix and message
func Encode(w io.Writer, addrs []uint8, msgID uint32, bodyFlag uint8, body []byte) (n int, err error) {
	var c int
	if c, err = EncodeRoute(w, addrs); err != nil {
		return
	}
	n += c
	if c, err = EncodeMsg(w, msgID, bodyFlag, body); err != nil {
		return
	}
	n += c
	return
}

// EncodeAsMsg encodes into new Msg
func EncodeAsMsg(addrs []uint8, msgID uint32, bodyFlag uint8, body []byte) (msg *Msg, err error) {
	var buf bytes.Buffer
	msg = &Msg{}
	msg.Head.MsgID = msgID
	msg.Head.Flag = Format
	msg.Head.BodyBytes = uint16(len(body) + 1)
	msg.Body.Flag = bodyFlag
	if addrs != nil {
		if _, err = EncodeRoute(&buf, addrs); err != nil {
			return
		}
	}
	if _, err = EncodeMsg(&buf, msgID, bodyFlag, body); err != nil {
		return
	}
	msg.Head.Raw = buf.Bytes()
	bodyOff := len(msg.Head.Raw) - int(msg.Head.BodyBytes)
	if addrs != nil {
		msg.Head.Prefix = msg.Head.Raw[0]
		msg.Head.Addrs = msg.Head.Raw[1 : len(addrs)+1]
		msg.Head.RawPrefix = msg.Head.Raw[0 : len(addrs)+1]
		msg.Head.RawHead = msg.Head.Raw[len(addrs)+1 : bodyOff]
	} else {
		msg.Head.RawHead = msg.Head.Raw[0:bodyOff]
	}
	msg.Body.Raw = msg.Head.Raw[bodyOff:]
	msg.Body.Data = msg.Body.Raw[1:]
	return
}

func encode7bit(w io.Writer, val uint64) (n int, err error) {
	data := make([]byte, 1)
	for {
		data[0] = uint8(val & 0x7f)
		val >>= 7
		if val != 0 {
			data[0] |= 0x80
		}
		if _, err = w.Write(data); err != nil {
			return
		}
		n++
		if val == 0 {
			break
		}
	}
	return
}

func decode7bit(w io.Writer, r io.Reader, maxBytes int) (val uint64, err error) {
	data := make([]byte, 1)
	for i := 0; i < maxBytes; i++ {
		if _, err = io.ReadFull(r, data); err != nil {
			return
		}
		if _, err = w.Write(data); err != nil {
			return
		}
		val |= (uint64(data[0]) & 0x7f) << (7 * uint(i))
		if (data[0] & 0x80) == 0 {
			break
		}
	}
	return
}

func decode7bit32(w io.Writer, r io.Reader) (uint32, error) {
	val, err := decode7bit(w, r, 4)
	return uint32(val), err
}

func decode7bit16(w io.Writer, r io.Reader) (uint16, error) {
	val, err := decode7bit(w, r, 2)
	return uint16(val), err
}
